test/AutoDiff/class_method.swift

```diff
@@ -0,0 +1,353 @@
+// RUN: %target-run-simple-swift
+// REQUIRES: executable_test
+
+import StdlibUnittest
+
+var ClassMethodTests = TestSuite("ClassMethods")
+
+ClassMethodTests.test("Final") {
+  final class Final : Differentiable {
+    func method(_ x: Float) -> Float {
+      return x * x
+    }
+  }
+
+  for i in -5...5 {
+    expectEqual(Float(i) * 2, gradient(at: Float(i)) { x in Final().method(x) })
+  }
+}
+
+ClassMethodTests.test("Simple") {
+  class Super {
+    @differentiable(wrt: x, jvp: jvpf, vjp: vjpf)
+    func f(_ x: Float) -> Float {
+      return 2 * x
+    }
+    final func jvpf(_ x: Float) -> (Float, (Float) -> Float) {
+      return (f(x), { v in 2 * v })
+    }
+    final func vjpf(_ x: Float) -> (Float, (Float) -> Float) {
+      return (f(x), { v in 2 * v })
+    }
+  }
+
+  class SubOverride : Super {
+    @differentiable(wrt: x)
+    override func f(_ x: Float) -> Float {
+      return 3 * x
+    }
+  }
+
+  class SubOverrideCustomDerivatives : Super {
+    @differentiable(wrt: x, jvp: jvpf2, vjp: vjpf2)
+    override func f(_ x: Float) -> Float {
+      return 3 * x
+    }
+    final func jvpf2(_ x: Float) -> (Float, (Float) -> Float) {
+      return (f(x), { v in 3 * v })
+    }
+    final func vjpf2(_ x: Float) -> (Float, (Float) -> Float) {
+      return (f(x), { v in 3 * v })
+    }
+  }
+
+  func classValueWithGradient(_ c: Super) -> (Float, Float) {
+    return valueWithGradient(at: 1) { c.f($0) }
+  }
+  expectEqual((2, 2), classValueWithGradient(Super()))
+  expectEqual((3, 3), classValueWithGradient(SubOverride()))
+  expectEqual((3, 3), classValueWithGradient(SubOverrideCustomDerivatives()))
+}
+
+ClassMethodTests.test("SimpleWrtSelf") {
+  class Super : Differentiable {
+    var base: Float
+    // FIXME(TF-648): Dummy to make `Super.AllDifferentiableVariables` be nontrivial.
+    var _nontrivial: [Float] = []
```

Comment:
Note that removing `_nontrivial` causes an assertion failure, as documented in [TF-648](https://bugs.swift.org/projects/TF/issues/TF-648):
```
SIL verification failed: vtable entry for #Super.f!1.jvp.1.S must be ABI-compatible
  ABI incompatible return values
  @convention(method) (@guaranteed Super) -> (Float, @owned @callee_guaranteed (@guaranteed Super.AllDifferentiableVariables) -> Float)
  @convention(method) (@guaranteed Sub) -> (Float, @owned @callee_guaranteed (Super.AllDifferentiableVariables) -> Float)
```

Detecting and diagnosing this case is difficult, so I decided to let this crash leak through for now. I'll work on the robust fix ([TF-625](https://bugs.swift.org/browse/TF-625)) soon.